tests/test_train.py

```diff
@@ -358,6 +363,258 @@ def test_train_nlu_autoconfig(
     assert args[1] == autoconfig.TrainingType.NLU
 
 
+def mock_async(monkeypatch, target, name):
+    mock = Mock()
+
+    async def mock_async_func(*args: Any, **kwargs: Any) -> None:
+        mock(*args, **kwargs)
+
+    monkeypatch.setattr(target, name, mock_async_func)
+    return mock
+
+
+def mock_core_training(monkeypatch):
+    return mock_async(monkeypatch, rasa.core, rasa.core.train.__name__)
+
+
+def mock_nlu_training(monkeypatch):
+    return mock_async(monkeypatch, rasa.nlu, rasa.nlu.train.__name__)
+
+
+def new_model_path_in_same_dir(old_model_path) -> str:
+    return str(Path(old_model_path).parent / (secrets.token_hex(8) + ".tar.gz"))
+
+
+class TestE2e:
+    def test_e2e_gives_experimental_warning(
+        self,
+        monkeypatch: MonkeyPatch,
+        trained_e2e_model: Text,
+        default_domain_path: Text,
+        default_stack_config: Text,
+        default_e2e_stories_file: Text,
+        default_nlu_data: Text,
+        caplog: LogCaptureFixture,
+    ):
+        mock_nlu_training(monkeypatch)
+        mock_core_training(monkeypatch)
+
+        with caplog.at_level(logging.WARNING):
+            train(
+                default_domain_path,
+                default_stack_config,
+                [default_e2e_stories_file, default_nlu_data],
+                output=new_model_path_in_same_dir(trained_e2e_model),
+            )
+
+        assert any(
+            [
+                "The end-to-end training is currently experimental" in record.message
+                for record in caplog.records
+            ]
+        )
+
+    def test_models_not_retrained_if_no_new_data(
+        self,
+        monkeypatch: MonkeyPatch,
+        trained_e2e_model: Text,
+        default_domain_path: Text,
+        default_stack_config: Text,
+        default_e2e_stories_file: Text,
+        default_nlu_data: Text,
+    ):
+        mocked_nlu_training = mock_nlu_training(monkeypatch)
+        mocked_core_training = mock_core_training(monkeypatch)
+
+        train(
+            default_domain_path,
+            default_stack_config,
+            [default_e2e_stories_file, default_nlu_data],
+            output=new_model_path_in_same_dir(trained_e2e_model),
+        )
+
+        mocked_core_training.assert_not_called()
+        mocked_nlu_training.assert_not_called()
+
+    def test_retrains_nlu_and_core_if_new_e2e_example(
+        self,
+        monkeypatch: MonkeyPatch,
+        trained_e2e_model: Text,
+        default_domain_path: Text,
+        default_stack_config: Text,
+        default_e2e_stories_file: Text,
+        default_nlu_data: Text,
+    ):
+        stories_yaml = rasa.shared.utils.io.read_yaml_file(default_e2e_stories_file)
+        stories_yaml["stories"][1]["steps"].append({"user": "new message!"})
+
+        new_stories_file = tempfile.NamedTemporaryFile(mode="w", suffix=".yml").name
```

Comment:
this won't clean up properly - can we use `tmp_path` instead?

Reply:
@joejuzl could you please take this over